p2p/protocol/holepunch/filter.go

```diff
@@ -0,0 +1,42 @@
+package holepunch
+
+import (
+	"github.com/libp2p/go-libp2p/core/peer"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// WithAddrFilter is a Service option that enables multi address filtering.
+// It allows to only allow a subset of observed addresses to the remote
+// peer. E.g., only announce TCP or QUIC multi addresses instead of both.
+// It also allows to only consider a subset of received multi addresses
+// that remote peers announced to us.
+// Theoretically, this API also allows to add multi addresses in both cases.
+func WithAddrFilter(maf AddrFilter) Option {
+	return func(hps *Service) error {
+		hps.filter = maf
+		return nil
+	}
+}
+
+// AddrFilter defines the interface for the multi address filtering.
+//   - FilterLocal is a function that filters the multi addresses that
+//     we send to the remote peer.
+//   - FilterRemote is a function that filters the multi addresses which
+//     we received from the remote peer.
+type AddrFilter interface {
+	FilterLocal(remoteID peer.ID, maddrs []ma.Multiaddr) []ma.Multiaddr
+	FilterRemote(remoteID peer.ID, maddrs []ma.Multiaddr) []ma.Multiaddr
+}
+
+// DefaultAddrFilter is the default address filtering logic. It strips
+// all relayed multi addresses from both the locally observed addresses
+// and received remote addresses.
+type DefaultAddrFilter struct{}
+
+func (d DefaultAddrFilter) FilterLocal(remoteID peer.ID, maddrs []ma.Multiaddr) []ma.Multiaddr {
+	return removeRelayAddrs(maddrs)
```

Comment:
Why are we not always removing the relay addresses? (Call `removeRelayAddrs` before calling to this interface).

Reply:
My reasoning was that the removal of relays from our own observed addresses is already a filtering step. I thought that people would be surprised to not have all multiaddresses at their disposal if they used that API.

Reply:
I can't think of a situation where having the relay address would be useful. Do you have any use case? If not, I'd advocate for just passing the relay-filtered addresses to the filter. Otherwise, we'd have to expose `removeRelayAddrs`.

Reply:
Sounds good, and done 👍